src/chunk.js

```diff
@@ -99,6 +119,183 @@ export function runChunksForTesting(nodeOrAmpDoc) {
   }
 }
 
+/**
+ * The priority of a chunk task. Higher priority tasks have higher values.
+ * @enum {number}
+ */
+export const ChunkPriority = {
+  HIGH: 20,
+  LOW: 10,
+  BACKGROUND: 0,
+};
+
+/** @enum {string} */
+const TaskState = {
+  NOT_RUN: 'not_run',
+  RUN: 'run',
+};
+
+/**
+ * A default chunkable task.
+ */
+class Task {
+  /**
+   * @param {!function(?IdleDeadline)} fn
+   */
+  constructor(fn) {
+    /** @public {TaskState} */
+    this.state = TaskState.NOT_RUN;
+
+    /** @private @const {!function(?IdleDeadline)} */
+    this.fn_ = fn;
+  }
+
+  /**
+   * Executes the wrapped function.
+   * @param {?IdleDeadline} idleDeadline
+   * @throws {Error}
+   */
+  runTask(idleDeadline) {
+    this.state = TaskState.RUN;
+    try {
+      this.fn_(idleDeadline);
+    } catch (e) {
+      this.onTaskError_(e);
+      throw e;
+    }
+  }
+
+  /**
+   * @return {string}
+   */
+  get name() {
+    return this.fn_.displayName || this.fn_.name;
+  }
+
+  /**
+   * Optional handling when a task run throws an error.
+   * @private
+   * @param {Error} unusedError
+   */
+  onTaskError_(unusedError) {
+    // By default, no-op.
+  }
+
+  /**
+   * Returns true if this task should be run without delay.
+   * @private
+   * @return {boolean}
+   */
+  immediateTriggerCondition_() {
+    // By default, there are no immediate trigger conditions.
+    return false;
+  }
+
+  /**
+   * Returns true if this task should be scheduled using `requestIdleCallback`.
+   * Otherwise, task is scheduled as macro-task on next event loop.
+   * @private
+   * @return {boolean}
+   */
+  useRequestIdleCallback_() {
+    // By default, always use requestIdleCallback.
+    return true;
+  }
+}
+
+/**
+ * A task that's run as part of AMP's startup sequence.
+ */
+class StartupTask extends Task {
+  /**
+   * @param {!function(?IdleDeadline)} fn
+   * @param {!Window} win
+   * @param {!./service/viewer-impl.Viewer|!Promise} viewerOrPromise
+   */
+  constructor(fn, win, viewerOrPromise) {
+    super(fn);
+
+    /** @private {!Window} */
+    this.win_ = win;
+
+    /** @private @const {boolean} */
+    this.active_ = isExperimentOnAllowUrlOverride(this.win_, 'chunked-amp');
+    if (!this.active_) {
+      return;
+    }
+
+    /** @private {?./service/viewer-impl.Viewer} */
+    this.viewer_ = null;
+
+    if (viewerOrPromise instanceof Promise) {
```

Comment:
Yes. `schedule_()` is called synchronously after a task is queued/executed, which calls `StartupTask#immediateTriggerCondition_()`, whose return value depends on the existence of `viewer_`.

We can change `schedule_()` to be executed as a micro-task instead, but I'd rather avoid behavioral changes in this already large-ish PR.